// @matterlabs/hardhat-zksync-solc/dist/src/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const task_names_1 = require("hardhat/builtin-tasks/task-names");
const config_env_1 = require("hardhat/internal/core/config/config-env");
require("./type-extensions");
const artifacts_1 = require("hardhat/internal/artifacts");
const compile_1 = require("./compile");
const utils_1 = require("./utils");
const child_process_1 = require("child_process");
const download_1 = require("hardhat/internal/util/download");
const fs_1 = __importDefault(require("fs"));
const ZK_ARTIFACT_FORMAT_VERSION = 'hh-zksolc-artifact-1';
const LATEST_VERSION = '1.2.0';
(0, config_env_1.extendConfig)((config, userConfig) => {
    var _a, _b, _c;
    const defaultConfig = {
        version: LATEST_VERSION,
        compilerSource: 'binary',
        settings: {
            compilerPath: '',
            experimental: {},
        },
    };
    if ((_b = (_a = userConfig === null || userConfig === void 0 ? void 0 : userConfig.zksolc) === null || _a === void 0 ? void 0 : _a.settings) === null || _b === void 0 ? void 0 : _b.optimizer) {
        console.warn('`optimizer` setting is deprecated, optimizer is always enabled');
    }
    config.zksolc = Object.assign(Object.assign({}, defaultConfig), userConfig === null || userConfig === void 0 ? void 0 : userConfig.zksolc);
    config.zksolc.settings = Object.assign(Object.assign({}, defaultConfig.settings), (_c = userConfig === null || userConfig === void 0 ? void 0 : userConfig.zksolc) === null || _c === void 0 ? void 0 : _c.settings);
});
(0, config_env_1.extendEnvironment)((hre) => {
    if (hre.network.config.zksync) {
        hre.network.zksync = hre.network.config.zksync;
        let artifactsPath = hre.config.paths.artifacts;
        if (!artifactsPath.endsWith('-zk')) {
            artifactsPath = artifactsPath + '-zk';
        }
        let cachePath = hre.config.paths.cache;
        if (!cachePath.endsWith('-zk')) {
            cachePath = cachePath + '-zk';
        }
        // Forcibly update the artifacts object.
        hre.config.paths.artifacts = artifactsPath;
        hre.config.paths.cache = cachePath;
        hre.artifacts = new artifacts_1.Artifacts(artifactsPath);
        // If solidity optimizer is not enabled, the libraries are not inlined and
        // we have to manually pass them into zksolc. That's why we force the optimization.
        hre.config.solidity.compilers.forEach((compiler) => {
            let settings = compiler.settings || {};
            compiler.settings = Object.assign(Object.assign({}, settings), { optimizer: { enabled: true } });
        });
    }
});
// This override is needed to invalidate cache when zksolc config is changed.
(0, config_env_1.subtask)(task_names_1.TASK_COMPILE_SOLIDITY_GET_COMPILATION_JOBS, async (args, hre, runSuper) => {
    const { jobs, errors } = await runSuper(args);
    jobs.forEach((job) => { job.solidityConfig.zksolc = hre.config.zksolc; });
    return { jobs, errors };
});
(0, config_env_1.subtask)(task_names_1.TASK_COMPILE_SOLIDITY_GET_ARTIFACT_FROM_COMPILATION_OUTPUT, async ({ sourceName, contractName, contractOutput, }, hre) => {
    var _a, _b, _c, _d;
    if (hre.network.zksync !== true) {
        return (0, artifacts_1.getArtifactFromContractOutput)(sourceName, contractName, contractOutput);
    }
    let bytecode = ((_b = (_a = contractOutput.evm) === null || _a === void 0 ? void 0 : _a.bytecode) === null || _b === void 0 ? void 0 : _b.object) || ((_d = (_c = contractOutput.evm) === null || _c === void 0 ? void 0 : _c.deployedBytecode) === null || _d === void 0 ? void 0 : _d.object) || '';
    bytecode = (0, utils_1.zeroxlify)(bytecode);
    let factoryDeps = {};
    let entries = Object.entries(contractOutput.factoryDependencies || {});
    for (const [hash, dependency] of entries) {
        factoryDeps[(0, utils_1.zeroxlify)(hash)] = dependency;
    }
    return {
        _format: ZK_ARTIFACT_FORMAT_VERSION,
        contractName,
        sourceName,
        abi: contractOutput.abi,
        // technically, zkEVM has no difference between bytecode & deployedBytecode,
        // but both fields are included just in case
        bytecode,
        deployedBytecode: bytecode,
        // zksolc does not support unlinked objects,
        // all external libraries are either linked during compilation or inlined
        linkReferences: {},
        deployedLinkReferences: {},
        // zkSync-specific field
        factoryDeps,
    };
});
(0, config_env_1.subtask)(task_names_1.TASK_COMPILE_SOLIDITY_RUN_SOLC, async (args, hre, runSuper) => {
    if (hre.network.zksync !== true) {
        return await runSuper(args);
    }
    if (hre.config.zksolc.settings.libraries) {
        args.input.settings.libraries = hre.config.zksolc.settings.libraries;
    }
    return await (0, compile_1.compile)(hre.config.zksolc, args.input, args.solcPath);
});
// This task is overriden to:
// - prevent unnecessary solc downloads when using docker
// - download zksolc binary if needed
// - validate zksolc binary
(0, config_env_1.subtask)(task_names_1.TASK_COMPILE_SOLIDITY_GET_SOLC_BUILD, async (args, hre, runSuper) => {
    var _a, _b;
    if (hre.network.zksync !== true) {
        return await runSuper(args);
    }
    if (hre.config.zksolc.compilerSource === 'docker') {
        // Versions are wrong here when using docker, because there is no
        // way to know them beforehand except to run the docker image, which
        // adds 5-10 seconds to startup time. We cannot read them from artifacts,
        // since that would make cache invalid every time, if the version is
        // different from the one in the docker image.
        //
        // If you wish to know the actual versions from build-info files,
        // please look at `output.version`, `output.long_version`
        // and `output.zk_version` in the generated JSON.
        return {
            compilerPath: '',
            isSolsJs: false,
            version: args.solcVersion,
            longVersion: '',
        };
    }
    const solcBuild = await runSuper(args);
    let compilerPath = hre.config.zksolc.settings.compilerPath;
    if (compilerPath) {
        const versionOutput = (0, child_process_1.spawnSync)(compilerPath, ['--version']);
        const version = (_b = (_a = versionOutput.stdout) === null || _a === void 0 ? void 0 : _a.toString().match(/\d+\.\d+\.\d+/)) === null || _b === void 0 ? void 0 : _b.toString();
        if (versionOutput.status !== 0 || version == null) {
            throw (0, utils_1.pluginError)(`Specified zksolc binary is not found or invalid`);
        }
    }
    else {
        compilerPath = await (0, utils_1.getZksolcPath)(hre.config.zksolc.version);
        if (!fs_1.default.existsSync(compilerPath)) {
            console.log(`Downloading zksolc ${hre.config.zksolc.version}`);
            try {
                await (0, download_1.download)((0, utils_1.getZksolcUrl)(hre.config.zksolc.version), compilerPath);
                fs_1.default.chmodSync(compilerPath, '755');
                console.log('Done.');
            }
            catch (e) {
                throw (0, utils_1.pluginError)(e.message.split('\n')[0]);
            }
        }
    }
    return solcBuild;
});
